arwen/contexts/metering.go

```diff
@@ -99,20 +105,204 @@ func (context *meteringContext) PopDiscard() {
 }
 
+func (context *meteringContext) PopMergeActiveState() {
+	stateStackLen := len(context.stateStack)
+	if stateStackLen == 0 {
+		return
+	}
+
+	prevState := context.stateStack[stateStackLen-1]
+	context.stateStack = context.stateStack[:stateStackLen-1]
+
+	context.initialGasProvided = prevState.initialGasProvided
+	context.initialCost = prevState.initialCost
+	context.gasForExecution = prevState.gasForExecution
+
+	context.addToGasUsedByAccounts(prevState.gasUsedByAccounts)
+}
+
+func (context *meteringContext) cloneGasUsedByAccounts() map[string]uint64 {
+	clone := make(map[string]uint64, len(context.gasUsedByAccounts))
+
+	for address, gasUsed := range context.gasUsedByAccounts {
+		clone[address] = gasUsed
+	}
+
+	return clone
+}
+
+func (context *meteringContext) addToGasUsedByAccounts(gasUsed map[string]uint64) {
+	for address, gas := range gasUsed {
+		context.gasUsedByAccounts[address] += gas
+	}
+}
+
+func (context *meteringContext) UpdateGasStateOnSuccess(vmOutput *vmcommon.VMOutput) error {
+	context.updateSCGasUsed()
+	context.setGasUsedToOutputAccounts(vmOutput)
+
+	err := context.checkGasLegacy()
+	if err != nil {
+		return err
+	}
+
+	err = context.checkGas(vmOutput)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (context *meteringContext) UpdateGasStateOnFailure(vmOutput *vmcommon.VMOutput) {
+	runtime := context.host.Runtime()
+	output := context.host.Output()
+
+	account, _ := output.GetOutputAccount(runtime.GetSCAddress())
+	account.GasUsed = math.AddUint64(account.GasUsed, context.GetGasProvided())
+}
+
+func (context *meteringContext) updateSCGasUsed() {
+	runtime := context.host.Runtime()
+	output := context.host.Output()
+
+	currentAccountAddress := runtime.GetSCAddress()
+	currentContractAccount, _ := output.GetOutputAccount(currentAccountAddress)
+	outputAccounts := context.host.Output().GetOutputAccounts()
+
+	gasTransferredByCurrentAccount := context.getGasTransferredByAccount(currentContractAccount)
+	gasUsedByOthers := context.getGasUsedByAllOtherAccounts(outputAccounts)
+
+	gasUsed := context.GasSpentByContract()
+	gasUsed = math.SubUint64(gasUsed, gasTransferredByCurrentAccount)
+	gasUsed = math.SubUint64(gasUsed, gasUsedByOthers)
+
+	context.gasUsedByAccounts[string(currentAccountAddress)] = gasUsed
+}
+
+func (context *meteringContext) TrackGasUsedByBuiltinFunction(err error, builtinInput *vmcommon.ContractCallInput, builtinOutput *vmcommon.VMOutput) error {
+	if err != nil {
+		context.UseGas(builtinInput.GasProvided)
+		return err
+	}
+
+	callerContractAddress := string(builtinInput.CallerAddr)
+	outputAccounts := builtinOutput.OutputAccounts
+	callerContractAccount, exists := outputAccounts[callerContractAddress]
+
+	gasTransferredByCurrentAccount := uint64(0)
+	if exists {
+		gasTransferredByCurrentAccount = context.getGasTransferredByAccount(callerContractAccount)
+	}
+
+	gasUsedByOthers := context.getGasUsedByAllOtherAccounts(outputAccounts)
```

Comment:
At the moment, this isn't actually needed, because the builtin functions that we currently have are not setting `outputAccount.GasUsed` in `VMOutput`. 

But if we ever have ideas about new builtin functions that will set `outputAccount.GasUsed`, the code is already here and doesn't need to be added.

Also, it's for consistency: builtin functions return a generic `VMOutput`, and this function treats it just like a generic `VMOutput`, without assumptions (including handling output accounts with `GasUsed`).